### PR TITLE
Fix race in event listener test

### PR DESCRIPTION
**Pull request by raghavsethi**
Changed TestEventListener to be single threaded. Also improved test to
re-use query runner and session.

### CODE REVIEW

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 1]
```diff
@@ -34,35 +36,40 @@
...
-            queryRunner = new DistributedQueryRunner(testSessionBuilder().build(), 1);
-            queryRunner.installPlugin(new TpchPlugin());
-            queryRunner.installPlugin(new TestingEventListenerPlugin(generatedEvents));
-            queryRunner.createCatalog("tpch", "tpch", ImmutableMap.of());
-        }
-        catch (Exception e) {
-            queryRunner.close();
-            throw e;
-        }
+        generatedEvents = new EventsBuilder();
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: Move this to the field initializer and make it final

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 2]
```diff
@@ -34,35 +36,40 @@
...
-        catch (Exception e) {
-            queryRunner.close();
-            throw e;
-        }
+        queryRunner = new DistributedQueryRunner(testSessionBuilder().build(), 1);
+        queryRunner.installPlugin(new TpchPlugin());
+        queryRunner.installPlugin(new TestingEventListenerPlugin(generatedEvents));
+        queryRunner.createCatalog("tpch", "tpch", ImmutableMap.of());
 
         session = testSessionBuilder()
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: This can also be final

**Comment by raghavsethi** [comment]
Finally ready for review :)

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 3]
```diff
@@ -70,61 +79,79 @@ private EventsBuilder generateEvents(@Language("SQL") String sql)
...
+            queryCompletedEvents = ImmutableList.builder();
+            splitCompletedEvents = ImmutableList.builder();
+
+            eventsLatch = new CountDownLatch(numEvents);
+        }
+
+        public void waitForEvents()
+                throws InterruptedException
+        {
+            eventsLatch.await();
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: This needs a timeout

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 4]
```diff
@@ -24,44 +24,53 @@
...
+    }
 
-        queryRunner.execute(session, sql);
+    @AfterClass(alwaysRun = true)
+    private void tearDown()
+    {
         queryRunner.close();
+    }
+
+    private synchronized EventsBuilder generateEvents(@Language("SQL") String sql, int numEventsExpected)
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: This doesn't need synchronized

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 5]
```diff
@@ -24,44 +24,53 @@
...
+            throws Exception
+    {
+        generatedEvents.initialize(numEventsExpected);
+        queryRunner.execute(session, sql);
+        generatedEvents.waitForEvents();
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: Pass in a timeout

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 6]
```diff
@@ -70,61 +79,79 @@ private EventsBuilder generateEvents(@Language("SQL") String sql)
...
+        EventsBuilder events = generateEvents("SELECT 1", 3);
 
-        assertEquals(events.getQueryCreatedEvents().size(), 1);
         QueryCreatedEvent queryCreatedEvent = getOnlyElement(events.getQueryCreatedEvents());
         assertEquals(queryCreatedEvent.getContext().getEnvironment(), "testing");
         assertEquals(queryCreatedEvent.getMetadata().getQuery(), "SELECT 1");
 
-        assertEquals(events.getQueryCompletedEvents().size(), 1);
         QueryCompletedEvent queryCompletedEvent = getOnlyElement(events.getQueryCompletedEvents());
         assertEquals(queryCompletedEvent.getStatistics().getTotalRows(), 0L);
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: Why zero rows? Shouldn't it be one row?

Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: Compare the query ID for the created and completed event

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 7]
```diff
@@ -24,44 +24,53 @@
...
-            queryRunner.createCatalog("tpch", "tpch", ImmutableMap.of());
-        }
-        catch (Exception e) {
-            queryRunner.close();
-            throw e;
-        }
+        queryRunner = new DistributedQueryRunner(testSessionBuilder().build(), 1);
+        queryRunner.installPlugin(new TpchPlugin());
+        queryRunner.installPlugin(new TestingEventListenerPlugin(generatedEvents));
+        queryRunner.createCatalog("tpch", "tpch", ImmutableMap.of("tpch.splits-per-node", "1"));
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: Make this an odd number like `3` so we get multiple split events

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 8]
```diff
@@ -70,61 +79,79 @@ private EventsBuilder generateEvents(@Language("SQL") String sql)
...
-        private final ImmutableList.Builder<QueryCreatedEvent> queryCreatedEvents = ImmutableList.builder();
-        private final ImmutableList.Builder<QueryCompletedEvent> queryCompletedEvents = ImmutableList.builder();
-        private final ImmutableList.Builder<SplitCompletedEvent> splitCompletedEvents = ImmutableList.builder();
+        private ImmutableList.Builder<QueryCreatedEvent> queryCreatedEvents;
+        private ImmutableList.Builder<QueryCompletedEvent> queryCompletedEvents;
+        private ImmutableList.Builder<SplitCompletedEvent> splitCompletedEvents;
+
+        private CountDownLatch eventsLatch;
+
+        public void initialize(int numEvents)
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java]: This also needs synchronized

**Comment by electrum** [comment]
Minor comments, otherwise looks good

**Pull request closed without merging**